Submit sample code for RGB to CMYK color conversion

larrylynn-wf opened the pull request:
Hello again Emmeran,

We were grateful for the work on 
https://github.com/rototor/pdfbox-graphics2d/issues/19 
and we wanted to contribute something back.

I saw in Issue # 6
https://github.com/rototor/pdfbox-graphics2d/issues/6 
that there was some interest in sample code for CMYK color mapping.  We wrote some code to do just that.  Our code has some proprietary entanglements, but it wasn't too much work to strip those out and make a generic version that could be released to the open source community without any licensing difficulties.  

The sample code we're offering is in RGBtoCMYKColorMapper.java.  The rest of the change is in testing to demonstrate that it works.

Adobe Preflight can be use to verify that this code (while somewhat rudimentary) is effective at transcoding vector images into the CMYK colorspace
<img width="1671" alt="2019-10-15_1137" src="https://user-images.githubusercontent.com/8104283/66855435-70beab00-ef40-11e9-9124-f9e1b2e96035.png">

I wasn't able to find any standard convention for where sample code belonged in a Java project.  But if you'd like to keep the sample code, it should be able to move around and maintain whatever directory hierarchy that you feel is appropriate.

larrylynn-wf commented:
Hello again,

I know a little bit about 'this whole color management mess' - enough to know that there is a whole lot more to it that I don't understand yet.  So thank you for the background information; that is useful.

In our codebase, we are using an ICC profile for RGB to CMYK conversion.  However, I believe our color profile is limited by copyright restrictions, so I stripped it out and used the naive approach instead when writing sample code.

I looked through http://www.eci.org/en/downloads.  I'm unable to find any information there about licensing and redistribution of the downloads.  If you can identify and ICC profile for CMYK conversion that is compatible with the Apache 2.0 license of pdfbox-graphics2d, I could rewrite this PR to do color conversion based on the ICC profile.  But since we don't know at the library level which printer would be used to print an output PDF, perhaps that's not useful either.

rototor commented:
Your class should always take a given ICC_Profile as parameter, because of course everybody needs another color profile. For tests you can use the ISO Coated V2 300 profile embedded in PDFBox. ```ICC_Profile.getInstance(PDDocument.class.getResourceAsStream("/org/apache/pdfbox/resources/icc/ISOcoated_v2_300_bas.icc"))``` should return the ICC_Profile for the embedded ISO Coated V2 profile. So PDFBox itself has already cleared this licensing question for us.

larrylynn-wf commented:
Hi,
I've refactored the sample code for RGB to CMYK conversion.  I backed out the naive approach and implemented a converter that relies on an ICC Profile and uses it to create an ICC Colorspace.  The constructor now takes an ICC Profile as a parameter.  In RenderSVGsTest.java I'm loading up  the copy of ISOcoated_v2_300_bas.icc that ships with PDFBox.

Is this a little closer to what you had in mind?

rototor commented on src/main/java/de/rototor/pdfbox/graphics2d/RGBtoCMYKColorMapper.java:
```diff
@@ -0,0 +1,58 @@
+    }
+
+    public void resetColorSpace(ICC_Profile icc_profile){
+        icc_colorspace = new ICC_ColorSpace(icc_profile);
```
You should construct a PDICCBased() based on this profile. i.e.
```
					this.pdProfile = new PDICCBased(document);
					OutputStream outputStream = pdProfile.getPDStream().createOutputStream(COSName.FLATE_DECODE);
					outputStream.write(profile.getData());
					outputStream.close();
					pdProfile.getPDStream().getCOSObject().setInt(COSName.N, 4);
					pdProfile.getPDStream().getCOSObject().setItem(COSName.ALTERNATE, COSName.DEVICECMYK);
```

To do so you of course need the PDDocument.

larrylynn-wf replied:
Updated.

rototor commented on src/main/java/de/rototor/pdfbox/graphics2d/RGBtoCMYKColorMapper.java:
```diff
@@ -0,0 +1,58 @@
+        float[] rgbFoats = rgbIntToFloat(rgbInts);
+        float[] cmykFloats = icc_colorspace.fromRGB(rgbFoats);
+
+        PDColor cmykColor = new PDColor(cmykFloats, PDDeviceCMYK.INSTANCE);
```
And here you should not use PDDeviceCMYK.INSTANCE, because thats wrong. You convert the colors using the profile but you don't tag the colors in the PDF as being in this colorspace.

Just use the ```pdProfile``` created in the comment above.

larrylynn-wf replied:
Updated.

larrylynn-wf commented:
Thank you for the feedback.  Updating now.

rototor commented:
Looks good to me. Thank you.

larrylynn-wf commented:
My pleasure.  Glad I could help.
